### PR TITLE
task: upgrade to quarkus 3.7.x

### LINKED ISSUE #26150: Upgrade to Quarkus 3.7.x
### Description

For the next version of Keycloak, we should upgrade Quarkus to be prepared to support the new LTS version of Quarkus.

### PR DESCRIPTION
**Pull request by shawkins**
closes: #26150

### CODE REVIEW

**Comment by shawkins** [comment]
@geoand we're failing to build due to the check on https://github.com/quarkusio/quarkus/blame/136b4d8ae9f19bfac8d4a13065085d3ff62dc19d/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java#L107

cc @vmuzikar

**Comment by shawkins** [comment]
@geoand @vmuzikar the migration away from resteasy-core seems like it will take some effort. The next commit adds a start at it, but it's incomplete. To use MockHttpRequest we still end up with a test dependency to resteasy-core. Then there's the ResteasyVertxProvider / KeycloakSessionContextInjector logic built up around the prior concept of the RestreasyContext - that doesn't seem to have a direct replacement in resteasy reactive. It could be that many of the calls to our org.keycloak.common.util.Resteasy class for context manipulation will need re-thought.

**Comment by geoand** [comment]
So essentially you need a way to turn off the Quarkus feature that  fails the build when the two stacks are combined?

**Comment by geoand** [comment]
> or is it technically possible to combine both stacks?

It is not really.

**Comment by vmuzikar** [comment]
@geoand We've been in some cases using it like that for the past few months. Can the explicit check be disabled in some way?

**Comment by geoand** [comment]
We can add a new `BuildItem that your extension would produce.

How does that sound for Keycloak?

**Comment by vmuzikar** [comment]
@geoand That might work but problem may be timing as we'd need this in the next Quarkus LTS.

But the way the check is currently done (checking if classic client is not present) not sure we can easily workaround it in Keycloak. @shawkins WDYT?

**Comment by geoand** [comment]
> That might work but problem may be timing as we'd need this in the next Quarkus LTS.

That should not be a problem, it's a very small change on our side :)

**Comment by geoand** [comment]
https://github.com/quarkusio/quarkus/pull/38214 is what I have in mind

**Comment by shawkins** [comment]
> But the way the check is currently done (checking if classic client is not present) not sure we can easily workaround it in Keycloak. @shawkins WDYT?

@vmuzikar the check that is failing is due to seeing resteasy-core, but not resteasy-client. It looks like we may be good wrt resteasy-client - it's only used in client side stuff, tests, and a couple of the adaptors. 

How far do you want to take the proactive removal resteasy-core - anything in this pr?, or just wait on the quarkus change an backout these changes for a separate issue?

**Comment by vmuzikar** [comment]
@shawkins Sorry, I misunderstood you meant [this](https://github.com/quarkusio/quarkus/blob/3.6.5/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java#L104-L108) check. But the main issue is actually [this](https://github.com/quarkusio/quarkus/blob/3.6.5/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java#L101-L103). Do I understand correctly now?

> How far do you want to take the proactive removal resteasy-core

Depends on the size of the changes, how confident are we about possible regressions. We have to consider we're relatively close to a release. My suggestion would be to go ahead with it as it needs to be done anyway at some point. We can always split it into a separate PR.

> or just wait on the quarkus change an backout these changes for a separate issue?

That'd be the cleanest solution for now (thanks @geoand for quickly providing the PR!) but not sure about the timing – we'd have to wait for a Quarkus release. We'd be blocked until then without knowing about other potential issues with Quarkus 3.6.

**Comment by shawkins** [comment]
> @shawkins Sorry, I misunderstood you meant [this](https://github.com/quarkusio/quarkus/blob/3.6.5/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java#L104-L108) check. But the main issue is actually [this](https://github.com/quarkusio/quarkus/blob/3.6.5/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java#L101-L103). Do I understand correctly now?

I do mean the former. I'm just trying to clarify it's because of our dependency on resteasy-core, not resteasy-client.

> Depends on the size of the changes, how confident are we about possible regressions. We have to consider we're relatively close to a release. My suggestion would be to go ahead with it as it needs to be done anyway at some point. We can always split it into a separate PR.

Ok, I'll leave in the parts that seem straight-forward and leave the context logic as is.

**Comment by vmuzikar** [comment]
@shawkins The fix from @geoand should be included in the released 3.7.0.CR1 Quarkus version. I think we can use it already instead of 3.6.

**Comment by shawkins** [comment]
@vmuzikar @geoand here are the next problems:

Currently keycloak manages jboss-logmanager to match what is in wildfly. Leaving that as-is fails because quarkus is now using the much newer 3.0.4, and has logic has references to classes only in that version - see LoggingResourceProcessor and ColorPatternFormatter. 

I think we can resolve this by removing the overrides, then anywhere we have a conflict with the wildfly and the quarkus versions, we need to add an explicit dependency to the quarkus version so it can pick up the transitive dependencies. But since we're moving up major versions, I don't know if there will be problems eventually uncovered from the wildfly side. If this gets straightened out (might take me another couple of trial commits) most of the failures should be resolved.

Another problem is that the bouncy castle version change that required several minor code updates may also be contributing to Fips UT classloading test failures - not sure yet what is going on there.

**Comment by geoand** [comment]
>         SRCFG00050: operator.name in EnvConfigSource does not map to any root
        SRCFG00050: operator.condition.name in EnvConfigSource does not map to any root

This is one likely for @radcortez

**Comment by vmuzikar** [comment]
@shawkins Thanks for the update.

> Currently keycloak manages jboss-logmanager to match what is in wildfly.

Not sure we still need to align with WF as the WF dist is gone now. Possibly it's because of some of our dependencies on wildfly-core? Or adapters? CC @keycloak/core.

**Comment by radcortez** [comment]
> The OLM test is failing because the operator deployment contains env properties for OPERATOR_NAME and OPERATOR_CONDITION_NAME, however we have Config interface that is mapping from the operator prefix - so there's an error at startup:
> 
> ```
>         SRCFG00050: operator.name in EnvConfigSource does not map to any root
>         SRCFG00050: operator.condition.name in EnvConfigSource does not map to any root
> ```
> 
> That seems to be new behavior for this release of quarkus.

No, this is not new behavior; this has been the default from mappings (except for the `quarkus` namespace). If you have a mapping on a specific namespace, every configuration available in the config system must be mapped in the mapping class.

Is this a new mapping, or are these properties new in the environment?

**Comment by shawkins** [comment]
> Is this a new mapping, or are these properties new in the environment?

@radcortez these are not new. The operator deployment is no different between 3.2 and 3.7, they both contain:

```
        - name: OPERATOR_NAME
          value: keycloak-operator
        - name: OPERATOR_CONDITION_NAME
          value: keycloak-operator.v86400000.0.0
```

Here's where we are directly adding the OPERATOR_NAME env value: https://github.com/keycloak/keycloak/blob/main/operator/scripts/create-olm-bundle.sh#L71

OPERATOR_CONDITION_NAME looks to be added by olm as it's not present in the deployment that's in the csv.

**Comment by shawkins** [comment]
@geoand @radcortez on the ShowConfigCommandTest.testShowConfigCommandHidesCredentialsInProfiles test failure - what is happening here is that we're running a series of tests in process, but need them to re-create the Config used by keycloak. I tried inserting a workaround similar to https://github.com/keycloak/keycloak/pull/26203/files#diff-61ddd5527dbeceadd6291d02a4aab62bf34223f602420e71a0182db2f7b39681R580 - but that didn't seem to work. If another test runs before that one, then we end up reusing some of the previous Config. What is the proper way fully reload the Config?

**Comment by shawkins** [comment]
The FipsDistTest failures are just due to needing the expected output to contain 1.000204 instead of 1.000203. 

The other fips failures I'm still working on. At least FipsUT runs fine locally, but of course that's not reproducing what's seen in the ci runs. Errors like this

```
2024-01-18T21:51:37.6895717Z [ERROR] org.keycloak.crypto.fips.test.FIPS1402HmacTest.testHmacSignaturesWithRandomSecretKey  Time elapsed: 0.001 s  <<< ERROR!
2024-01-18T21:51:37.6897446Z java.lang.NoClassDefFoundError: Could not initialize class java.util.UUID$Holder
2024-01-18T21:51:37.6898506Z 	at java.base/java.util.UUID.randomUUID(UUID.java:148)
2024-01-18T21:51:37.6899702Z 	at org.keycloak.jose.HmacTest.testHmacSignaturesWithRandomSecretKey(HmacTest.java:49)
```

have been seen with bc before - [https://github.com/bcgit/bc-java/issues/1149 - ](https://github.com/bcgit/bc-java/issues/1149#issuecomment-1790457576) - that specifically mentions trying 1.0.2.4 with keycloak.

Should I just try downgrading to 1.0.2.3, or is anyone else already familiar with bc classloading issues? cc @mposolda @mabartos

**Comment by mabartos** [comment]
@shawkins AFAIK, the only recent thing related to the FIPS was this[1], which resolves these failures[2] with `KcSamlEncryptedIdTest`. But I don't know anything about the BC versions. @mposolda and @rmartinc know more about it.

[1] https://github.com/keycloak/keycloak/pull/26303
[2] https://github.com/shawkins/keycloak/actions/runs/7595105517/job/20687484590#step:6:2255

**Comment by shawkins** [comment]
> @shawkins AFAIK, the only recent thing related to the FIPS was this[1], which resolves these failures[2] with `KcSamlEncryptedIdTest`. But I don't know anything about the BC versions. @mposolda and @rmartinc know more about it.
> 
> [1] #26303 [2] https://github.com/shawkins/keycloak/actions/runs/7595105517/job/20687484590#step:6:2255

Thanks @mabartos yes that looks like it wasn't included yet. I've rebased to pick that up and backed up the change to the FipsDistTest expection - https://github.com/keycloak/keycloak/commit/d01624d602341c62bd69b6c2571895ebb28523ad - that should run almost clean.

**Comment by vmuzikar** [comment]
@shawkins Thanks for the summary.

I think we can keep the BC version for now (i.e. override what comes from Quarkus) and upgrade as a follow-up.

Do you have any potential solutions for the ConfigCommandTest and OLM tests?

**Comment by shawkins** [comment]
> Do you have any potential solutions for the ConfigCommandTest and OLM tests?

Unfortunately not yet. The direct solution for the OLM issue is unfortunately using a different prefix for our Operator Config class (which we closed as an issue not that long ago). The ConfigCommandTest I'll make a few more stabs at, but it seems like we're going to need some proper guidance on how to fully re-initialize the smallrye config.

**Comment by vmuzikar** [comment]
> The direct solution for the OLM issue is unfortunately using a different prefix for our Operator Config

We could do that. We don't document the config options anywhere, should be ok-ish to rename as we consider it an internal one.

**Comment by shawkins** [comment]
> We could do that.

We started down that path here https://github.com/keycloak/keycloak/pull/24931 - do you want to revive that and account for anything that may be trying to directly manipulate the image value?

**Comment by radcortez** [comment]
> @geoand @radcortez on the ShowConfigCommandTest.testShowConfigCommandHidesCredentialsInProfiles test failure - what is happening here is that we're running a series of tests in process, but need them to re-create the Config used by keycloak. I tried inserting a workaround similar to https://github.com/keycloak/keycloak/pull/26203/files#diff-61ddd5527dbeceadd6291d02a4aab62bf34223f602420e71a0182db2f7b39681R580 - but that didn't seem to work. If another test runs before that one, then we end up reusing some of the previous Config. What is the proper way fully reload the Config?

So you mean that the configuration names that are making the test fail are required for some other test, and they leak to a test that is not required?

**Comment by shawkins** [comment]
> So you mean that the configuration names that are making the test fail are required for some other test, and they leak to a test that is not required?

If you run the failing test by itself, it will pass. If a test without a config file runs before the failing test with a config file, then the latter will fail - because it won't see the options from the config file. Whatever used to be reinitializing the full configuration in between tests is no longer working. The simple change from ConfigurationTest does not work here - there is still somewhere else that we're picking up the previous config from.

**Comment by shawkins** [comment]
> Seems like the cleanest solution, I'm afraid... We'll have to update any impacted processes (test pipeline, release pipelines) on our end.

@vmuzikar @radcortez to recap - it sounds like this only worked before due to luck. That is if we have a Config interface with a configmapping prefix of operator, then the presense of env values like OPERATOR_NAME, which do not match our interface should cause an exception. The proper fix is to reopen #12352, but just to get unblocked in the meantime I've added dummy methods to Config interface.

**Comment by shawkins** [comment]
> If you run the failing test by itself, it will pass. If a test without a config file runs before the failing test with a config file, then the latter will fail - because it won't see the options from the config file. Whatever used to be reinitializing the full configuration in between tests is no longer working. The simple change from ConfigurationTest does not work here - there is still somewhere else that we're picking up the previous config from.

@radcortez if I printout the config instance and the TCCL when the ShowConfig command runs in each test the output is identical:

io.smallrye.config.SmallRyeConfig@553d9329 QuarkusClassLoader:Quarkus Runtime ClassLoader: TEST for testShowConfigCommandShowsRuntimeConfig(LaunchResult) (QuarkusTest) restart no:0@39b50027

So we're definitely bleeding over the config / classloader from the previous test method into another one. 

We have some additional integration test logic related to the config with TestKeycloakConfigSourceProvider - but as best as I can tell that is not involved here. There also appears to be quarkus test logic that is responsible for ConfigProviderResolver.setInstance(null) - but as I saw with ConfigurationTest that seems to be insufficient as there is caching  going on behind the scenes per class loader.

cc @pedroigor

**Comment by radcortez** [comment]
> So we're definitely bleeding over the config / classloader from the previous test method into another one.

This is expected. Quarkus starts / stops per class test and not per method test. When a different configuration is required to run a test that conflicts with other tests in the test class, I recommend creating a separate test class with a different configuration.

**Comment by shawkins** [comment]
> This is expected. Quarkus starts / stops per class test and not per method test. When a different configuration is required to run a test that conflicts with other tests in the test class, I recommend creating a separate test class with a different configuration.

Since we have quite a few CLITests with multiple methods, it won't be great splitting each of those into a separate class. Would it be okay to do something like https://github.com/keycloak/keycloak/pull/26203/commits/0f08756c2809a73c2691634579b3a51e5438c618 - so that we can reinitialize the config via a test resource?

**Comment by radcortez** [comment]
> Since we have quite a few CLITests with multiple methods, it won't be great splitting each of those into a separate class. Would it be okay to do something like [0f08756](https://github.com/keycloak/keycloak/commit/0f08756c2809a73c2691634579b3a51e5438c618) - so that we can reinitialize the config via a test resource?

It could work if you are only interested in the CLI Config. A configuration change between tests will not affect Quarkus.

An alternative could be to disable the unmapped validation using `quarkus.config.mapping.validate-unknown=false`. This is global, but may be acceptable for tests. A safer alternative is to add a [SmallRyeConfigBuilderCustomizer](https://smallrye.io/smallrye-config/3.5.2/config/customizer/) only in the testing module and call `withMappingIgnore` to ignore only the `operator` namespace.

File: pom.xml
Comment (mabartos) [inline pom.xml:49]:
```suggestion
        <quarkus.version>3.7.0</quarkus.version>
        <quarkus.build.version>3.7.0</quarkus.build.version>
```
AFAIK, should be ready now.

**Review by mabartos — COMMENTED** [review]
@shawkins Thanks! In general, LGTM. Just some minor things. 

Don't we also override some deps from quarkus-bom? I see that we have defined some Hibernate props separately (`hibernate.c3p0.version`, `hibernate-orm.plugin.version`) with the lower version of Hibernate. AFAIK, the first one should be for testing, and the other one - I don't think it's relevant anymore. 

My point is, whether we are confident that we updated all relevant dependencies or not. 

Moreover, as we're touching the `crypto` stuff, it'd be great to have the "approved" stamp from the @keycloak/core .

File: pom.xml
Comment (shawkins) [reply pom.xml:49]: On the versions I'm not sure about alignment. I looked for things in the quarkus runtime dependency tree that were managed and tried to update appropriately.

Now I see there is a script in set-quarkus-version, which I've updated to consider all dependency properties. This pulls in a few more alignments - but I had to also remove a couple of child pom overrides[*], one downgrade, and the osgi version change to 6.0 did not seem to work either. And of course if our property names do not line up with what is in quarkus, there will be things missed here as well. The current overlap (from the output of the script) is:

Setting 2.2.224 to h2
Setting 14.0.21.Final to infinispan
Setting 2.0.1 to jakarta.mail
Setting 6.2.7.Final to resteasy
Setting 2.0.6 to slf4j
Setting 2.2.3.Final to wildfly-elytron
Setting 4.25.1 to liquibase
Setting 6.0.0 to osgi
Setting 8.0.33 to mysql-jdbc
Setting 42.7.1 to postgresql-jdbc
Setting 3.3.2 to mariadb-jdbc
Setting 12.4.2.jre11 to mssql-jdbc
Setting 23.3.0.23.09 to oracle-jdbc
Setting 2.2 to hamcrest
Setting 4.2.0 to awaitility
Setting 1.19.3 to testcontainers

- I switched the version and dependency name  around in the pr for better readability.

Just like with the logging changes that ar proposed here, we can also consider removing some of the version overrides that we have to align better with quarkus - then re-introduce them only if we need to customize the version.

On using 3.7.0 - I didn't determine the reason, but trying that version locally doens't work for me - maven can't access the pom. 

* I'm not so sure about how something like the resteasy legacy version is currently working. It seems to miss that the resteasy spi is a transitive dependency as well.

Comment (vmuzikar) [reply pom.xml:49]: With regards to the deps versions alignment. IIRC, it should not be needed to do manually, nor using the script. @Pepo48 Did some work around it to remove most of the overrides (to implicitly align with quarkus-bom) and keep only those strictly needed for a reason.

Comment (vmuzikar) [reply pom.xml:49]: From a quick look, seems the explicitly stated versions are mostly for testsuites.

Comment (shawkins) [reply pom.xml:49]: @vmuzikar yes, some of those are for tests. A breakdown of what I think this looks like:

Versions where we trail quarkus - we will spawn follow-on issues to upgrade:
- bc-fips to 1.0.2.4 
- liquibase to 4.25.1
- osgi to 6.0

Versions where we are ahead of quarkus, and that are used for more than tests - might be worth a quarkus issue for them to upgrade:
- jakarata.mail

Versions where we are trail quarkus, and that are used for more than tests, but seem to upgrade without an issue:
- infinispan
- resteasy
- slf4j
- wildfly-elytron

We can consider removing our version overrides for these - but resteasy and wildfly-elytron have some specfic overrides though that need accounted for in child poms.

The rest are used for tests, and updating to the later version associated with quarkus seems fine.

Comment (vmuzikar) [reply pom.xml:49]: > We can consider removing our version overrides for these - but resteasy and wildfly-elytron have some specfic overrides though that need accounted for in child poms. We can also split these upgrade task off to separate issues if that seems better.

Yes, I'd address it as a follow-up.

Infinispan should probably not be upgraded with Quarkus as per https://github.com/keycloak/keycloak/issues/21363#issuecomment-1629117378 (CC @keycloak/store)

Comment (shawkins) [reply pom.xml:49]: The additional version alignment will be deffered to https://github.com/keycloak/keycloak/issues/26535

Comment (shawkins) [reply pom.xml:49]: To circle back to @mabartos 3.7.0 still does not work for me locally:

```
[ERROR]   The project org.keycloak:keycloak-parent:999.0.0-SNAPSHOT (/home/shawkins/workspaces/keycloak/keycloak/pom.xml) has 1 error
[ERROR]     Non-resolvable import POM: The following artifacts could not be resolved: io.quarkus.platform:quarkus-bom:pom:3.7.0 (absent): io.quarkus.platform:quarkus-bom:pom:3.7.0 was not found in https://maven.repository.redhat.com/techpreview/all/ during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of redhat-techpreview-all-repository has elapsed or updates are forced @ line 355, column 25 -> [Help 2]
```

And a similar failure happened with a gha build:

```
2024-01-26T17:39:44.2977248Z ##[error]Unable to process file command 'output' successfully.
2024-01-26T17:39:44.2980367Z ##[error]Invalid format '[ERROR] Non-resolvable import POM: Could not find artifact io.quarkus.platform:quarkus-bom:pom:3.7.0 in central (https://repo.maven.apache.org/maven2) @ line 355, column 25'
```

cc @geoand

Comment (geoand) [reply pom.xml:49]: Until the platform is released (next week), you need to use `io.quarkus` as the group id

Comment (shawkins) [reply pom.xml:49]: Thanks @geoand I'll make the update once that happens.

Comment (shawkins) [reply pom.xml:49]: @vmuzikar as far as I can tell the basic follow-up task of alignment would look like https://github.com/shawkins/keycloak/compare/iss26150...shawkins:keycloak:iss26535 - from there we would still need to tackle those upgrades.

Comment (vmuzikar) [reply pom.xml:49]: > Until the platform is released (next week), you need to use `io.quarkus` as the group id

Yeah, let's use CR1 for now until Platform is released.

> as far as I can tell the basic follow-up task of alignment would look like

Thanks! Can you please create an issue for it? Probably belongs under this epic: https://github.com/keycloak/keycloak/issues/21356

**Review by vmuzikar — DISMISSED** [review]
LGTM, nice job @shawkins!

File: pom.xml
Comment (ahus1) [reply pom.xml:49]: Comment about the override of the Infinispan version: Upgrade it to a later version if Quarkus already includes it. Still, keep the separate BOM in the root pom.xml as in the previous months we've been tracking usually a more recent version of Infinispan with the latest fixes we needed, and IMHO this will continue at least for KC24. 

If with the latest Quarkus upgrade we're upgrading to a newer version of Infinispan and h2, the explicit mention of those dependencies can vanish from the root pom.xml in this or a following PR.

Comment (shawkins) [reply pom.xml:49]: > Thanks! Can you please create an issue for it? Probably belongs under this epic: https://github.com/keycloak/keycloak/issues/21356

Yes that commit is already targeted toward #26535

> Comment about the override of the Infinispan version: Upgrade it to a later version if Quarkus already includes it.

I mispoke about infinispan being an upgrade - it's an overlapping dependency, but we're already even with quarkus. The corrected categorization of dependency alignment is on #26535

**Review by mabartos — DISMISSED** [review]
@shawkins Ok, let's use the CR version for now and do the additional stuff in the follow-up task. 

@keycloak/core Could you please review the changes around the BouncyCastle?

**Review by abstractj — DISMISSED** [review]

**Comment by mabartos** [comment]
@shawkins jfyi; conflict in the operator module

**Comment by shawkins** [comment]
> @shawkins jfyi; conflict in the operator module

Sure that's bound to happen the longer this waits. I'll address that and any thing else that comes up before we're ready to merge.

**Comment by mabartos** [comment]
We'll try to ping guys from the @keycloak/core team :P

File: crypto/default/src/main/java/org/keycloak/crypto/def/BCOCSPProvider.java
Comment (mposolda) [inline crypto/default/src/main/java/org/keycloak/crypto/def/BCOCSPProvider.java:440]: Minor: I think `org.bouncycastle.asn1.DERIA5String` can be removed from imports in this class?

File: crypto/default/src/main/java/org/keycloak/crypto/def/BCUserIdentityExtractorProvider.java
Comment (mposolda) [inline crypto/default/src/main/java/org/keycloak/crypto/def/BCUserIdentityExtractorProvider.java:165]: Minor: looks that `DERUTF8String` can be removed from the imports of this class?

File: pom.xml
Comment (mposolda) [inline pom.xml:90]: Question: did you tried the other way around?

I mean instead of overriding `bouncycastle.bcfips.version` with the lower version than quarkus bom, the alternative might be to update`bouncycastle.pkixfips.version` to newer version?

If it doesn't work, is it possible to create follow-up issue for look into this? It can be nice to use the latest BCFIPS libraries to avoid risk of having some security issues...

**Review by mposolda — COMMENTED** [review]
@shawkins I've added 2 very minor inline comments for imports (feel free to ignore them if there are no other changes planned for this PR).

But one question regarding BCFIPS versions.

Also I am not sure if this part of the docs will need some update regarding version - https://github.com/keycloak/keycloak/blob/main/docs/guides/server/fips.adoc#bouncycastle-fips-bits (unfortunately those versions are still hardcoded in the docs instead of being dynamically retrieved, which means we need to manually update docs when the pom versions are updated).

File: crypto/default/src/main/java/org/keycloak/crypto/def/BCOCSPProvider.java
Comment (shawkins) [reply crypto/default/src/main/java/org/keycloak/crypto/def/BCOCSPProvider.java:440]: It cannot, it's still used by DERIA5String.getInstance

File: pom.xml
Comment (shawkins) [reply pom.xml:90]: @mposolda there are no newer versions of pkixfips.

> If it doesn't work, is it possible to create follow-up issue for look into this?

See #26468 

> It can be nice to use the latest BCFIPS libraries to avoid risk of having some security issues...

Yes, it looks like a vulnerability was reported against 1.0.2.3

**Comment by shawkins** [comment]
Resolved the conflict and squashed the commits. Unfortunately that means some approvals need to be re-added.

**Review by vmuzikar — DISMISSED** [review]
@shawkins Thanks for the rebase. AFAICT, there are no real changes after the rebase. IMHO we don't need all the re-approvals.

**Comment by shawkins** [comment]
> Thanks for the rebase. AFAICT, there are no real changes after the rebase. IMHO we don't need all the re-approvals.

Sounds good. Needed to tweak the conflict resoultion, so with just another approval the auto-merge should do it.

**Review by vmuzikar — APPROVED** [review]

**Comment by vmuzikar** [comment]
@mposolda Do you want to give this a final look? :)

**Comment by vmuzikar** [comment]
Reopening as this had to be temporarily reverted here: https://github.com/keycloak/keycloak/pull/26681

**Pull request merged**
